src/Avalonia.Controls/Expander.cs

```diff
@@ -332,5 +288,63 @@ private void UpdatePseudoClasses()
 
             PseudoClasses.Set(":expanded", IsExpanded);
         }
+
+        /// <summary>
+        /// Called when the <see cref="IsExpanded"/> property has to be coerced.
+        /// </summary>
+        /// <param name="value">The value to coerce.</param>
+        protected virtual bool OnCoerceIsExpanded(bool value)
+        {
+            RoutedEventArgs eventArgs;
+
+            if (value)
+            {
+                eventArgs = new RoutedEventArgs(ExpandingEvent, this);
+                OnExpanding(eventArgs);
+            }
+            else
+            {
+                eventArgs = new RoutedEventArgs(CollapsingEvent, this);
+                OnCollapsing(eventArgs);
+            }
+
+            if (eventArgs.Handled)
+            {
+                // If the event was externally handled (canceled) we must still notify the value has changed.
+                // This property changed notification will update any external code observing this property that itself may have set the new value.
+                // We are essentially reverted any external state change along with ignoring the IsExpanded property set.
+                // Remember IsExpanded is usually controlled by a ToggleButton in the control theme and is also used for animations.
+                _ignorePropertyChanged = true;
+
+                RaisePropertyChanged(
+                    IsExpandedProperty,
+                    oldValue: value,
+                    newValue: !value,
+                    BindingPriority.LocalValue,
+                    isEffectiveValue: true);
```

Comment:
I think everyone agrees that this is non-ideal and kind-of 'hacky'. It was first discussed here: https://github.com/AvaloniaUI/Avalonia/pull/9555#discussion_r1035415374. 

This is why I was hoping coercion would handle this automatically for us but unfortunately it doesn't. Therefore, I had to add this back (you can remove it and test the control catalog to see the issues). I think WPF has issues here as well but for sure it would be better to improve coercion and actually handle the real property value that comes back after coercion itself.

For now, it must be done to ensure external code state is correct. (It is essentially just used to notify of a property changed event and force updates) Keep in mind the expander itself isn't the main 'source' of the IsExpanded property -- it is actually the checked state of a toggle button in the control template... so we have to be careful about this.

> The problem is that when two-way bindings transfer a new value in either direction, they assume that it was accepted by the recipient.

Yes, that's the real issue. One I think exists in WPF as well.

> the new value has local priority 

That's something to think about some more I suppose. I copy pasted the code and when it was a direct-property LocalValue made sense. Now perhaps it doesn't.

Reply:
> This is why I was hoping coercion would handle this automatically for us but unfortunately it doesn't.

Yeah I think we should probably make coercion handle this if it's not too difficult. What doesn't work with coercion?

Reply:
> What doesn't work with coercion?

Here was a nice summary:

>If the recipient object changes the value to something other than the one that was set, a property changed event is raised and the binding can update the sender object. But if the recipient object instead ignores the new value (as is the case here), then no event is raised and no action is taken.

Basically when a cancel happens in the coerce method the property value is reverted back to its original value. This means the value hasn't actually changed and no change notification happens -- seems to make sense. However, if two controls are two-way bound this actually causes a desync. Because the sender will still change it's state.

My simple understanding (without looking at any control flow) is:
1. Sender changes value (say true -> false)
2. Receiver coerces value
3. Coercion raises cancel events and external code cancels the change (value switch false -> true)
4. The value 'true' is set to the control... but this is the same as it was before. Therefore no change notification and updates.
5. The sender still has a false value instead of true because it never got a change notification because the value didn't actually change.

So an additional check is needed to see if coercion changes the value to a value other than what when into the coerce method. If so, and there is a two-way binding, it needs to notify the sender of a change even when the value set to the control is actually the same.

Reply:
True, and the ViewModel case is the issue I've seen in WPF before if I recall correctly. So this is an old problem.

Reply:
@grokys  In the example I gave:

 * sender : This is the IsChecked property of the ToggleButton within the control template. 
 * receiver : The actual IsExpanded property of the Expander control itself

This is a general problem though with two-way binding if the "receiver" doesn't accept the sender's value for whatever reason.

Reply:
Ok, thanks for the clarification, that makes sense.

The problem with the `RaisePropertyChanged` hack here is that this method is `internal` so the hack will only work within Avalonia itself. We probably need to work out how to do this properly.

Reply:
Supporting cancellation in events is a fairly common practice from my experience and the API is mostly standardized as discussed above. It really should be supported in a UI framework IMO. Flyout is another example. If it currently requires an internal-use-only work-around for now that seems justifiable to me (especially a harmless one). The issue with the property system has existed for almost 20 years.

It's common in some cases to need to raise a property changed notification "hack" in apps as well. Especially due to the view model example above. It might also be possible to promote this idea from a "hack" into a "feature" by allowing a clean way to force a refresh.